File: contracts/generator/game-clock.ts
/**
 * Golden vectors for the game clock the player shows: `M:SS`, or `H:MM:SS`
 * from the first hour on, fractions floored, and a bad value read as `0:00`.
 */
import { DEFAULT_TOLERANCE, vectorCase, type VectorFile } from "./vector";

import { formatGameClock } from "@/features/player/format-timecode";

function clockCase(name: string, totalSeconds: number) {
  return vectorCase(name, "formatGameClock", { totalSeconds }, (i) =>
    formatGameClock(i.totalSeconds),
  );
}

export function buildGameClock(): VectorFile {
  return {
    contract: "game-clock",
    description:
      "formatGameClock writes a game time in seconds as M:SS, or H:MM:SS once " +
      "the game reaches an hour. Fractions of a second are floored, and a " +
      "negative time reads 0:00 (non-finite input too, which JSON cannot carry).",
    reference: ["src/features/player/format-timecode.ts"],
    tolerance: DEFAULT_TOLERANCE,
    cases: [
      clockCase("the game start", 0),
      clockCase("under a minute", 7),
      clockCase("a fraction is floored", 59.999),
      clockCase("one minute", 60),
      clockCase("minutes keep no padding", 754.5),
      clockCase("just under an hour", 3599.98),
      clockCase("the first hour", 3600),
      clockCase("minutes are padded after an hour", 3725.02),
      clockCase("a long game", 36000),
      clockCase("a negative time", -2),
    ],
  };
}

File: mac/HockeyKit/Sources/HockeyCore/GameClock.swift
/// The game clock the player shows.
///
/// Port of `formatGameClock` in `src/features/player/format-timecode.ts`,
/// pinned by `contracts/vectors/game-clock.json`.

/// A game time in seconds as `M:SS`, or `H:MM:SS` from the first hour on.
/// Fractions are floored; a negative or non-finite time reads `0:00`.
public func formatGameClock(_ totalSeconds: Double) -> String {
    // The cap only keeps an absurd value from trapping the integer conversion.
    let safe = totalSeconds.isFinite && totalSeconds > 0
        ? Int(min(totalSeconds, 1e15).rounded(.down)) : 0
    let hours = safe / 3600
    let minutes = (safe % 3600) / 60
    let seconds = safe % 60
    let paddedSeconds = seconds < 10 ? "0\(seconds)" : "\(seconds)"
    if hours > 0 {
        let paddedMinutes = minutes < 10 ? "0\(minutes)" : "\(minutes)"
        return "\(hours):\(paddedMinutes):\(paddedSeconds)"
    }
    return "\(minutes):\(paddedSeconds)"
}
